samples/nrf9160/serial_lte_modem/src/main.c

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <logging/log.h>
+LOG_MODULE_REGISTER(app, CONFIG_THIN_LOG_LEVEL);
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <uart.h>
+#include <string.h>
+#include <net/socket.h>
+#include <init.h>
+#include <bsd.h>
+#include <lte_lc.h>
+#include <net/mqtt.h>
+
+#include <hal/nrf_gpio.h>
+#include <hal/nrf_power.h>
+#include <hal/nrf_regulators.h>
+
+#include "th_test.h"
+#include "th_connect.h"
+#if defined(CONFIG_LTE_LINK_CONTROL)
+#include "th_modem_service.h"
+#endif
+#if defined(CONFIG_THIN_MQTT_SERVICE)
+#include "th_mqtt_service.h"
+#endif
+#if defined(CONFIG_THIN_LWM2M_SERVICE)
+#include "th_lwm2m_service.h"
+#endif
+#if defined(CONFIG_THIN_GPS_SERVICE)
+#include "th_gps_service.h"
+#endif
+#if defined(CONFIG_THIN_SNTP_SERVICE)
+#include "th_sntp_service.h"
+#endif
+#if defined(CONFIG_THIN_TCPIP_SERVICE)
+#include "th_tcpip_service.h"
+#endif
+
+#define INVALID_DESCRIPTOR	-1
+
+#define AT_BUF_SIZE		CONFIG_INTER_CONNECT_UART_BUF_SIZE
+#define THREAD_STACK_SIZE	1048
+#define THREAD_PRIORITY		K_PRIO_PREEMPT(CONFIG_THIN_MODEM_THREAD_PRIO)
+
+static int at_socket_fd = INVALID_DESCRIPTOR;
+static struct pollfd fds[1];
+static int nfds;
+static u8_t at_buf[AT_BUF_SIZE];
+static size_t at_buf_len;
+static struct k_work at_cmd_send_work;
+static struct k_thread socket_thread;
+static K_THREAD_STACK_DEFINE(socket_thread_stack, THREAD_STACK_SIZE);
+static bool at_cmd_responded;
+
+static void at_cmd_send(struct k_work *work)
+{
+	int bytes_sent;
+
+	ARG_UNUSED(work);
+
+	bytes_sent = send(at_socket_fd, at_buf, at_buf_len, 0);
+
+	if (bytes_sent <= 0) {
+		LOG_WRN("Could not send AT command to modem: %d", bytes_sent);
+	}
+}
+
+static void at_socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int err;
+	int r_bytes;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	while (1) {
+		/* Poll the socket for incoming data. */
+		err = poll(fds, nfds, K_FOREVER);
+		if (err < 0) {
+			LOG_ERR("Poll error: %d", err);
+			continue;
+		}
+
+		/* Read AT socket in non-blocking mode. */
+		r_bytes = recv(at_socket_fd, at_buf, AT_BUF_SIZE - 4,
```

Comment:
Why `AT_BUF_SIZE - 4` and not simply `AT_BUF_SIZE`?

Reply:
4 is the reserved size for protocol overhead (STX, TYPE, LENGTH, BCC). Actually it should be 5, plus the Notification type as the buffer is reserved for receive.